src/memory_region.rs

```diff
@@ -33,52 +27,89 @@ use std::{
 
 /// Callback executed before generate_access_violation()
-pub type MemoryCowCallback = Box<dyn Fn(u32) -> Result<u64, ()>>;
+///
+/// Safety: Reentrancy is forbidden.
+pub type AccessViolationHandler = Box<dyn Fn(&mut MemoryRegion, u64, AccessType, u64, u64)>;
 /// Fail always
 #[allow(clippy::result_unit_err)]
-pub fn default_memory_cow_callback(_cow_callback_payload: u32) -> Result<u64, ()> {
-    Err(())
+pub fn default_access_violation_handler(
+    _region: &mut MemoryRegion,
+    _region_max_len: u64,
+    _access_type: AccessType,
+    _vm_addr: u64,
+    _len: u64,
+) {
+}
+macro_rules! access_violation_guard {
+    ($self_ty:ty, $self:expr, $access_type:expr, $vm_addr:expr, $len:expr) => {{
+        if let Some((index, region)) = $self.find_region($vm_addr) {
+            if let Some(host_addr) = region.vm_to_host($access_type, $vm_addr, $len) {
+                return ProgramResult::Ok(host_addr);
+            }
+            let mut region = (*region).clone();
+            let max_len = $self
+                .regions
+                .get(index.saturating_add(1))
+                .map_or(u64::MAX, |next_region| next_region.vm_addr)
+                .saturating_sub(region.vm_addr);
+            {
+                // Safety: The RefCell prevents reentrancy in AccessViolationHandler.
+                let access_violation_handler = $self.access_violation_handler.borrow_mut();
+                (&access_violation_handler)(&mut region, max_len, $access_type, $vm_addr, $len);
+            }
+            if let Err(err) = $self.replace_region(index, region) {
+                return ProgramResult::Err(err);
+            }
+            if let Some((_index, region)) = $self.find_region($vm_addr) {
+                if let Some(host_addr) = region.vm_to_host($access_type, $vm_addr, $len) {
+                    return ProgramResult::Ok(host_addr);
+                }
+            }
+        }
+        generate_access_violation(
+            $self.config,
+            $self.sbpf_version,
+            $access_type,
+            $vm_addr,
+            $len,
+        )
+    }};
 }
 
 /// Memory region for bounds checking and address translation
-#[derive(Default, Eq, PartialEq)]
+#[derive(Default, Eq, PartialEq, Clone)]
 #[repr(C, align(32))]
 pub struct MemoryRegion {
     /// start host address
-    pub host_addr: Cell<u64>,
+    pub host_addr: u64,
     /// start virtual address
     pub vm_addr: u64,
-    /// end virtual address
-    pub vm_addr_end: u64,
     /// Length in bytes
     pub len: u64,
     /// Size of regular gaps as bit shift (63 means this region is continuous)
     pub vm_gap_shift: u8,
     /// Is `AccessType::Store` allowed without triggering an access violation
-    pub writable: Cell<bool>,
-    /// User defined payload for the [MemoryCowCallback]
-    pub cow_callback_payload: u32,
+    pub writable: bool,
+    /// User defined payload for the [AccessViolationHandler]
+    pub access_violation_handler_payload: Option<u16>,
```

Comment:
This payload isn't used anywhere. Is it necessary?

Reply:
It can be read from the `region` parameter of the `AccessViolationHandler`. See `test_access_violation_handler_region_id()` for example.